describe @exportS3Method for non-owned generics and delayed method registration (#1087)
hadley/r-pkgs: dependencies-in-practice.Rmd

  UseMethod("count")
}

> #' @exportS3method

Reviewer: I think that this can still be `@export`? I'm not sure I'm yet ready to say that every S3 method should use `@exportS3method`.

Author: ```suggestion
#' @export
```

...
```

Now imagine that your package implements a method for `count()` for a class you "own" (not `data.frame`) but a generic that you "don't own".
A good example is the dbplyr package, which implements dplyr's `count()` generic for dbplyr's `tbl_lazy` class.

In this case, `@export` will not work because it assumes the `count()` generic is included or imported in the dbplyr NAMESPACE, so instead we need to use `@exportS3Method`, providing the precise generic that we're providing a method for.

```{r}
#| eval: false
#' @exportS3Method dplyr::count
count.tbl_lazy <- function(x, ..., wt = NULL, sort = FALSE, name = NULL) { ... }
```

In `NAMESPACE`, we have:

```         

> S3method(count,tbl_lazy)

Reviewer: Ooh this will change too.

Author: Whoops, yep

Now imagine that your package implements a method for `count()` for a class you "own" (not `data.frame`).
A good example is the dbplyr package, which implements `count()` for the `tbl_lazy` class.

> In this case, `@export` will not work because it assumes the `count()` generic is included or imported in the dbplyr NAMESPACE, so instead we need to use `@exportS3Method`.

Reviewer: ```suggestion
In this case, `@export` will not work because it assumes the `count()` generic is included or imported in the dbplyr NAMESPACE, so instead we need to use `@exportS3Method`, providing the precse generic that we're providing a method for.
```

> In this case, `@export` will not work because it assumes the `count()` generic is included or imported in the dbplyr NAMESPACE, so instead we need to use `@exportS3Method`, providing the precse generic that we're providing a method for.

Author: ```suggestion
In this case, `@export` will not work because it assumes the `count()` generic is included or imported in the dbplyr NAMESPACE, so instead we need to use `@exportS3Method`, providing the precise generic that we're providing a method for.
```

> This will use "conditional" method registration, which means the method will only be registered when the dplyr package is loaded.

Reviewer: I think it would be worth a sentence or two here (or maybe below) to also mention that this also works even if the package is only suggested.

Author: Updated with a sentence below (and removes mention of "conditional" method registration for import dependencies)

```{r}
#| eval: false
#' @exportS3Method dplyr::count
count.tbl_lazy <- function(x, ..., wt = NULL, sort = FALSE, name = NULL) { ... }
```

In `NAMESPACE`, we have:

```         
S3method(count,tbl_lazy)
```

> dbplyr also provides methods for various generics provided by the base package, such as `dim()` and `names()` - in this case, we can use `@export` or `@exportS3Method base::names`

Reviewer: I think I'd still recommend using `@export` here because we don't need conditional registration.

Author: ```suggestion
dbplyr also provides methods for various generics provided by the base package, such as `dim()` and `names()` - in this case, we can still use `@export` since we don't need conditional registration. 
```

  c(NA, length(op_vars(x$lazy_query)))
}

> #' @exportS3Method base::names

Author: ```suggestion
#' @export
```

```         
S3method(dim,tbl_lazy)
...

> S3method(base::names,tbl_lazy)

Author: ```suggestion
S3method(names,tbl_lazy)
```
